### PR TITLE
gh-119740: Remove deprecated trunc delegation

### PR DESCRIPTION
**Pull request by mdickinson**
This PR removes the delegation of `int` to the `__trunc__` special method: `int` will now only delegate to `__int__` and `__index__` (in that order). `__trunc__` continues to exist, but its sole purpose is to support `math.trunc`.

* Issue: gh-119740

----
📚 Documentation preview 📚: https://cpython-previews--119743.org.readthedocs.build/

### CODE REVIEW

File: Doc/library/functions.rst
Comment (serhiy-storchaka) [inline Doc/library/functions.rst]: It should be `versionchanged`. `deprecated` should be used for the deprecation note and now removed.

File: Doc/reference/datamodel.rst
Comment (serhiy-storchaka) [inline Doc/reference/datamodel.rst]: The same as above.

File: Doc/library/functions.rst
Comment (serhiy-storchaka) [inline Doc/library/functions.rst]: Perhaps ``` :func:`int` ``` would be better for this. `int` is documented as a function and a class. In this case we refer to a callable.

File: Doc/whatsnew/3.14.rst
Comment (serhiy-storchaka) [inline Doc/whatsnew/3.14.rst]: I think that it should be either ```The :func:`int` built-in``` or ```The :class:`int` constructor```. Preferably the former, since we will get a link to the constructor.

File: Doc/library/functions.rst
Comment (mdickinson) [reply Doc/library/functions.rst]: Okay, so "versionchanged:: 3.14" for the "int will no longer delegate message", right?

I don't understand the "and now removed" bit. Are you saying that the earlier 3.11 versionchanged note should be removed? Or that it should be changed to say "deprecated"? Or it should stay as it was?

File: Doc/whatsnew/3.14.rst
Comment (mdickinson) [reply Doc/whatsnew/3.14.rst]: Done in https://github.com/python/cpython/pull/119743/commits/7492da8cb64922147227858e9ef5cd120632a8ba

File: Doc/whatsnew/3.14.rst
Comment (serhiy-storchaka) [inline Doc/whatsnew/3.14.rst]:
```suggestion
  :func:`!int` must implement either :meth:`~object.__int__` or
```
To avoid repeating the same link within a paragraph.

File: Doc/library/functions.rst
Comment (serhiy-storchaka) [reply Doc/library/functions.rst]: I think that `deprecated:: 3.11` should be used instead of `versionchanged:: 3.11` when the deprecation started. Now it is removed, so all is fine.

**Review by serhiy-storchaka — APPROVED** [review]
LGTM. But I would prefer to merge this after #119687, to make backporting of #119687 easier.

File: Doc/whatsnew/3.14.rst
Comment (serhiy-storchaka) [reply Doc/whatsnew/3.14.rst]: Wait. it should be
```suggestion
  :class:`int` must implement either :meth:`~object.__int__` or
```
or
```suggestion
  :class:`!int` must implement either :meth:`~object.__int__` or
```
or simply
```suggestion
  integer must implement either :meth:`~object.__int__` or
```
We refer now to `int` as a type, not a callable.

File: Doc/whatsnew/3.14.rst
Comment (picnixz) [inline Doc/whatsnew/3.14.rst]: NIT (for consistency):

```suggestion
  :meth:`~object.__index__`. (Contributed by Mark Dickinson in :gh:`119743`.)
```

File: Doc/library/functions.rst
Comment (picnixz) [inline Doc/library/functions.rst]: Just a personal opinion, but I think 

```suggestion
      :func:`int` no longer delegates to the :meth:`~object.__trunc__` method.
```

would be clearer.

**Comment by mdickinson** [comment]
Thanks for the comments; I'll address before merging (but it may take me a few days to get to it).

**Comment by mdickinson** [comment]
> But I would prefer to merge this after #119687, to make backporting of #119687 easier.

Definitely! Looks like #119687 has gone in now; I'll update for conflicts when I get the chance.

File: Doc/whatsnew/3.14.rst
Comment (mdickinson) [inline Doc/whatsnew/3.14.rst]:
```suggestion
* The :func:`int` built-in no longer delegates to
```

File: Doc/reference/datamodel.rst
Comment (mdickinson) [inline Doc/reference/datamodel.rst]:
```suggestion
      :func:`int` no longer delegates to the :meth:`~object.__trunc__` method.
```

File: Doc/library/functions.rst
Comment (mdickinson) [reply Doc/library/functions.rst]: Thanks; that makes sense. I wasn't sure whether we're supposed to remove older `deprecated` notes when things were finally removed.

**Review by serhiy-storchaka — APPROVED** [review]

**Comment by mdickinson** [comment]
Updating for merge conflicts ...

**Pull request merged**
